src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -146,12 +151,21 @@ long getThrottlingWaitTimeMicros() {
 
     CompletableFuture<Integer> subscribe(String topic, QualityOfService qos) {
         return connect().thenCompose((b) -> {
-            logger.atDebug().kv(TOPIC_KEY, topic).kv(QOS_KEY, qos.name()).log("Subscribing to topic");
+            logger.atDebug().kv(TOPIC_KEY, topic).kv(QOS_KEY, qos.name())
+                    .log("Subscribing to topic");
             synchronized (this) {
                 throwIfNoConnection();
-                return connection.subscribe(topic, qos).thenApply((i) -> {
-                    subscriptionTopics.put(topic, qos);
-                    return i;
+                inprogressSubscriptions.incrementAndGet();
+                return connection.subscribe(topic, qos).whenComplete((i, error) -> {
+                    if (error == null) {
+                        subscriptionTopics.put(topic, qos);
+                        logger.atDebug().kv(TOPIC_KEY, topic).kv(QOS_KEY, qos.name())
+                                .log("Successfully subscribed to topic");
+                    } else {
+                        logger.atError().kv(TOPIC_KEY, topic)
+                                .cause(error).log("Error subscribing to topic");
+                    }
+                    inprogressSubscriptions.decrementAndGet();
```

Comment:
synchronize this block (in whenComplete) with `this` too so that subscriptionTopics and `inprogressSubscription` reading _and_ writing is synchronized so that our counting is correct.